pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -86,90 +101,370 @@ func (c *ClusterMemberController) sync(ctx context.Context, syncCtx factory.Sync
+// no member is found to scale up return empty string.
+func (c *ClusterMemberController) getEtcdPeerURLToScaleUp(ctx context.Context) (string, error) {
+	network, err := c.networkLister.Get("cluster")
+	if err != nil {
+		return "", fmt.Errorf("failed to list cluster network: %w", err)
+	}
+
+	members, err := c.etcdClient.MemberList(ctx)
 	if err != nil {
-		return nil, err
+		return "", err
 	}
+	memberMap, _ := c.getMemberMap(ctx, members)
 
-	// go through the list of all pods, pick one peerFQDN to return from unready pods
-	// and collect dns resolution errors on the way.
-	for _, pod := range pods {
-		if !strings.HasPrefix(pod.Name, "etcd-") {
+	nodes, err := c.nodeLister.List(labels.SelectorFromSet(labels.Set{masterLabel: ""}))
+	if err != nil {
+		return "", err
+	}
+
+	// Sort nodes by created timestamp.
+	sort.Slice(nodes, func(i, j int) bool {
+		return nodes[i].CreationTimestamp.Before(&nodes[j].CreationTimestamp)
+	})
+
+	// Get the oldest voting member's timestamp
+	var oldestMemberTimestamp *metav1.Time
+	for _, node := range nodes {
+		if _, isMember := memberMap[node.Name]; isMember {
+			oldestMemberTimestamp = &node.CreationTimestamp
+		}
+	}
+	if oldestMemberTimestamp.IsZero() {
+		return "", fmt.Errorf("could not find any node mapped to a member")
+	}
+
+	// Only scale up if current membership is less than the max allowed
+	currentMemberCount := len(members)
+	maxMemberCount := c.replicaCount + 1 // Surge up
+	if currentMemberCount >= maxMemberCount {
+		klog.Warningf("Skipping scale up etcd membership at maximum size: %d", maxMemberCount)
+		return "", nil
+	}
+
+	for _, node := range nodes {
+		// Ignore if the node is mapped to a member and already part of the quorum.
+		_, isMember := memberMap[node.Name]
+		if isMember {
+			klog.V(4).Infof("node/%s is already mapped to an etcd member", node.Name)
 			continue
 		}
-		isEtcdContainerRunning, isEtcdContainerReady := false, false
-		for _, containerStatus := range pod.Status.ContainerStatuses {
-			if containerStatus.Name != "etcd" {
+
+		// To avoid re-adding a recently scaled-down/member-less node back into the cluster
+		// filter out any nodes older than the current oldest member node.
+		// We assume those older nodes to already be scaled-down.
+		if node.CreationTimestamp.Before(oldestMemberTimestamp) {
+			// TODO(haseeb): What happens if the node is the same age as the oldest member node?
+			// Could be a recently removed node or it could be the next new node we need to scale up (e.g 3 new nodes created at the same time)
+			// In practice the nodes seem to differ in their CreationTimestamps but there should be a better way to mark scaled-down nodes.
+			//
+			// Is it safe to assume that the to-be-scaled-down nodes are always deleted first?
+			// If so, the Machine deletion timestamp could be used.
+			//
+			// Or a new Node/Machine annotation or condition added by the etcd-operator on scaled-down nodes.
+			continue
```

Comment:
@hexfusion Could use your thoughts around how best to identify/tag previously scaled-down nodes.
The quorum-protection proposal [assumes](https://github.com/openshift/enhancements/pull/943/files#diff-8f08f89622ea85c6276166feaa7be109adaae9cf5d280b18ef20f4f218fce87cR161) that the replaced machines will always be deleted first although I'm not sure that will always be the case.
```
A Control Plane Machine is marked for deletion and a new Control Plane Machine is created
```
But if so then I can just use the deletion timestamp instead of inferring scaled-down nodes by their age.

Reply:
Thanks a lot for the extensive review!
I'll circle back to the other comments later but you're right, I don't have a valid scenario in mind for when you would want a Machine/Node to be replaced but not deleted.

The implementation should be a lot simpler to reason about if I can assume a non-nil deletion timestamp on Machines to be scaled down so I'm going to update the logic here accordingly. I've already uncovered an embarrassing number of bugs in this PR due to the whole dance around timestamps and inferring scaled down members.